pkg/relayer/session/sessiontree.go

```diff
@@ -0,0 +1,190 @@
+) (SessionTree, error) {
+	// join the storePrefix and the session.sessionId to create a unique storePath
+	storePath := filepath.Join(storesDirectory, session.SessionId)
+
+	treeStore, err := smt.NewKVStore(storePath)
+	if err != nil {
+		return nil, err
+	}
+
+	// create the SMST from the KVStore and a nil value hasher so the proof would contain a non-hashed Relay
+	// that could be used to validate the proof on-chain.
+	tree := smt.NewSparseMerkleSumTree(treeStore, sha256.New(), smt.WithValueHasher(nil))
+
+	sessionTree := &sessionTree{
+		session:   session,
+		storePath: storePath,
+		treeStore: treeStore,
+		tree:      tree,
+
+		removeFromRelayerSessions: removeFromRelayerSessions,
+	}
+
+	return sessionTree, nil
+}
+
+// GetSession returns the session corresponding to the SMST.
+func (st *sessionTree) GetSession() *sessiontypes.Session {
+	return st.session
+}
+
+// Update is a wrapper for the SMST's Update function. It updates the SMST with the given key, value, and weight.
+// This function should be called by the Miner when a Relay has been successfully served.
+// It returns an error if the SMST has been flushed to disk which indicates that updates are no longer allowed.
+func (st *sessionTree) Update(key, value []byte, weight uint64) error {
+	st.sessionMu.Lock()
+	defer st.sessionMu.Unlock()
+
+	if st.claimedRoot != nil {
+		return ErrSessionTreeClosed
+	}
+
+	return st.tree.Update(key, value, weight)
+}
+
+// ProveClosest is a wrapper for the SMST's ProveClosest function. It returns a proof for the given path.
+// This function should be called when a session has been claimed and needs to be proven.
+// If the proof has already been generated, it returns the cached proof.
+// It returns an error if the SMST has not been flushed yet (the claim has not been generated)
+func (st *sessionTree) ProveClosest(path []byte) (proof *smt.SparseMerkleClosestProof, err error) {
+	st.sessionMu.Lock()
+	defer st.sessionMu.Unlock()
+
+	// A claim need to be generated before a proof can be generated.
+	if st.claimedRoot == nil {
+		return nil, ErrSessionTreeNotClosed
+	}
+
+	// If the proof has already been generated, return the cached proof.
+	if st.proof != nil {
+		return st.proof, nil
+	}
+
+	// Restore the KVStore from disk since it has been closed after the claim has been generated.
+	st.treeStore, err = smt.NewKVStore(st.storePath)
+	if err != nil {
+		return nil, err
+	}
+
+	st.tree = smt.ImportSparseMerkleSumTree(st.treeStore, sha256.New(), st.claimedRoot, smt.WithValueHasher(nil))
+
+	// Generate the proof and cache it along with the path for which it was generated.
+	st.proof, err = st.tree.ProveClosest(path)
+	st.proofPath = path
+
+	return st.proof, err
+}
+
+// Flush gets the root hash of the SMST needed for submitting the claim; then commits the entire tree to disk
+// and stops the KVStore.
+// It should be called before submitting the claim on-chain. This function frees up the in-memory resources.
```

Comment:
Are you sure about `This function frees up the in-memory resources.`.

I know we're assigning some vars to to nil, but isn't it up to the GC?

Reply:
It should free up the KVStore resources (will update the comment)